catroid/src/main/java/org/catrobat/catroid/stage/HmsSpeechRecognitionHolder.kt

```diff
@@ -0,0 +1,121 @@
+/*
+ * Catroid: An on-device visual programming system for Android devices
+ * Copyright (C) 2010-2021 The Catrobat Team
+ * (<http://developer.catrobat.org/credits>)
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * An additional term exception under section 7 of the GNU Affero
+ * General Public License, version 3, is available at
+ * http://developer.catrobat.org/license_additional_term
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package org.catrobat.catroid.stage
+
+import android.content.Context
+import android.content.Intent
+import android.os.Bundle
+import android.speech.RecognizerIntent
+import android.util.Log
+import com.huawei.hms.mlplugin.asr.MLAsrCaptureConstants
+import com.huawei.hms.mlsdk.asr.MLAsrConstants
+import com.huawei.hms.mlsdk.asr.MLAsrListener
+import com.huawei.hms.mlsdk.asr.MLAsrRecognizer
+import org.catrobat.catroid.formulaeditor.SensorHandler
+import org.catrobat.catroid.utils.ToastUtil
+import java.lang.ref.WeakReference
+
+class HmsSpeechRecognitionHolder : SpeechRecognitionHolderInterface {
+    companion object {
+        private val TAG = this.toString()
+    }
+
+    private var speechRecognizer: MLAsrRecognizer? = null
+    private var speechIntent: Intent? = null
+    private var context = WeakReference<Context>(null)
+
+    private val listener = object : MLAsrListener {
+        override fun onResults(result: Bundle?) {
+            val recognizedString = result?.getString(MLAsrRecognizer.RESULTS_RECOGNIZED)
+            Log.d(TAG, "onResults: $recognizedString")
+            callback?.onResult(recognizedString.orEmpty())
+        }
+
+        override fun onRecognizingResults(result: Bundle?) = Unit
+
+        override fun onError(error: Int, errorMessage: String?) {
+            Log.d(TAG, "$error $errorMessage")
+            when (error) {
+                MLAsrConstants.ERR_NO_NETWORK ->
+                    // TODO error message
+                    ToastUtil.showError(context.get(), "No network")
```

Comment:
Please remove the comment and replace the string with a resource string so that it can be translated.